### PR TITLE
Feature/customizeable dataset

### PR DESCRIPTION
**Pull request by SublimeProphets**
Hi Daniel,

I've invested the whole day extending your plugin with some additional features.

**main added features**
- Added new fields
  - permalink
  - tags
  - categories
  - long_description
  - excerpt
  - stock_quantity,
  - stock_status
- Extend Backend
  - introduced new section for basic fields with simple checkboxes to enable or disable
- Added Logic for attributes
  - toggle to enable / disable attributes
  - Handling for the visibility boolean
  - Handling for the variation boolean
  - Whitelisting based on all existing WooCommerce attributes
  - Numeric interpolation (probably not the most used feature tho but had to do it for my client anyway)
- Fixed some bugs
  -  the `verify_settings_nonce` was inactive, `update_settings_options` has been called on every adminpage whatsoever.
  - downgraded requirements from PHP 8.1 down to PHP 8.0 due it is working perfectly well on PHP 8.0 (and WP 6.1.1 is yet not fully PHP 8.1 compatible)
- extended the readme slightly

**tested on**
- a quiet big WC instance with dozens of combinations (all latest versions
- clean fresh install (PHP 8.0, all latest versions)

**whats missing or could be further improved?**
- Didn't found a fast and simple way to extend the .pot file wihout paying licence fees (and loco translate doesn't recognize the new string). If you provide me an updated .pot file, I am happy to provide some nice human readable strings for the basic fields.
- screenshot could be updated after string translations
- the data structure of tags and categories is simply the label for now, could be also customizable to the needs - may some want the slug, description or so......

I hope i could fix the Codacy errors, was not able to set it up in a senseful timespan. if there are problems with it, please provide me some instruction :-) 

Greets Chris

### CODE REVIEW

**Review by w3bdesign — CHANGES_REQUESTED** [review]
I am attaching the output from the code quality scanners

**Comment by w3bdesign** [comment]
@SublimeProphets 

image

image

**Comment by SublimeProphets** [comment]
Hi @w3bdesign 
I've made several changes to reduce code complexity, duplication and so on and managed to resolve most of the issues.
However, I am not sure how i should proceed with reducing the complexity here.
Would love to have some advice :-)

### [Codacy](https://app.codacy.com/manual/w3bdesign/algolia-woo-indexer/commit?cid=698095960) detected an issue:
#### Message: `The method get_product_attributes() has an NPath complexity of 8804. The configured NPath complexity threshold is 200.`

    
#### Currently on:
+ **Commit**:  37c16d9990352bad7da04c4d01eba4aae96ae5bf
+ **File**:    [classes/class-send-products.php](https://github.com/w3bdesign/algolia-woo-indexer/blob/37c16d9990352bad7da04c4d01eba4aae96ae5bf/classes/class-send-products.php)
+ **LineNum**: [244](https://github.com/w3bdesign/algolia-woo-indexer/blob/37c16d9990352bad7da04c4d01eba4aae96ae5bf/classes/class-send-products.php#L244)

**Comment by w3bdesign** [comment]
@SublimeProphets 

https://modess.io/npath-complexity-cyclomatic-complexity-explained

https://en.wikipedia.org/wiki/Cyclomatic_complexity

I am no Npath expert, but it seems like your code is too complex?

**Comment by w3bdesign** [comment]
@SublimeProphets 

I haven't analyzed the code in great detail, but my initial guess is that you should do something about the foreach loop there.

File: classes/class-send-products.php [hunk 1]
```diff
@@ -105,6 +177,153 @@ public static function get_product_type_price($product)
...
+            $setting_ids = explode(",", $setting_ids);
+            $setting_ids_interp = get_option(ALGOWOO_DB_OPTION . ATTRIBUTES_INTERP);
+            $setting_ids_interp = explode(",", $setting_ids_interp);
+
+            if (!$rawAttributes) {
+                return false;
+            }
+
+            $attributes = [];
+            foreach ($rawAttributes as $attribute) {
```
Comment (w3bdesign) [inline classes/class-send-products.php:270]: @SublimeProphets 

I guess it is this line?

**Review by w3bdesign — CHANGES_REQUESTED** [review]
@SublimeProphets 

This foreach should be refactored

**Pull request closed without merging**